### ISSUE TITLE
zetta-device onRemoteUpdate / onRemoteFetch hooks

### ISSUE THREAD

**Issue by AdamMagaluk** [zettajs/zetta#215]
- Provide way to limit properties that are sent via the API and z2z.
- Provide way to verify and limit properties that are updated via the API.

**Comment by AdamMagaluk** [comment]
Currently looking at coming up with a design for the remote update api. I have a few options i'm looking at below. @mdobson Any thoughts on this.

```js
// Use prototype and manually call save. Can return an error in the callback.
Device.prototype.onRemoteUpdate = function(properties, done) {
  var self = this;

  var reserved = ['macAddress', 'ipAddress'];
  Object.keys(properties).forEach(function(key) {
    // filter out reserved properties
    if (reserved.indexOf(key) !== -1) {
      return;
    }
    
    self[key] = properties[key];
  });

  this.save(done);

  // or

  done(new Error('Dont update that.'));
};

// Use prototype but don't manually call save. Just return a filtered object. Not async. Can return a error.
Device.prototype.onRemoteUpdate = function(properties) {
  var self = this;

  var reserved = ['macAddress', 'ipAddress'];
  Object.keys(properties).forEach(function(key) {
    // filter out reserved properties
    if (reserved.indexOf(key) !== -1) {
      delete properties[key];
    }
  });

  return properties;
  // or
  return new Error('Dont update that.');
};

// Option to add remote update function to config object.
Device.prototype.init = function(config) {
  var self = this;

  config
    .type('light')
    .remoteUpdate(function(properties, done) {
      // Could use either of the methods above or others. "this" scoping gets kinda weird.
      var reserved = ['macAddress', 'ipAddress'];
      Object.keys(properties).forEach(function(key) {
        // filter out reserved properties
        if (reserved.indexOf(key) !== -1) {
          return;
        }
        self[key] = properties[key];
      });
      
      self.save(done);
    })
};
```

**Comment by mdobson** [comment]
Through email #3 looks most consistent with how we structure functionality
currently. Also, with this approach we could attach a function to the
prototype instead of an inline calls back. Or am Would that be tougher to
do?

**Comment by AdamMagaluk** [comment]
Yeah 3 definitely feels more zettaish. I also had a thought about our current filtering of properties to not update monitor values and zetta reserved properties. I think we also have to by default filter out all prototype names.

"Also, with this approach we could attach a function to the prototype instead of an inline calls back" Not sure i'm understanding this.

**Comment by mdobson** [comment]
I probably used inline wrong.

Inline

.map("foo", function() {});

Prototype

.map("foo", this.foo);

+1 on the filtering.

-Matt

**Comment by AdamMagaluk** [comment]
Yeah i'm liking option 3 even more. You could use either method of inlining or assigning a prototype function just like .map supports.

I'm thinking the internal of option 1 where if you use a remoteUpdate you would have to update properties yourself, call save to do a db save, and fire the cb when complete.

**Comment by AdamMagaluk** [comment]
Here is what i'm thinking for the reverse side of that. Remote fetch using the config method. Just return an object for what the properties in the api should be.

```js
Device.prototype.init = function(config) {
  config
    .type('light')
    .remoteFetch(function() {
      return {}; // Return whatever you want.
    })
};
```

Should `id`, `type`, and `name` be overridden by zetta? Should leading underscores be filtered out by zetta?

**Comment by mdobson** [comment]
Stellar. Looks good to me. I would say yes to both of your questions as
well. Those sets of properties should be protected.

**Comment by AdamMagaluk** [comment]
Implemented in #224
